mkdocs_redirects/plugin.py

```diff
@@ -68,55 +100,146 @@ def get_html_path(path, use_directory_urls):
     return f.dest_path.replace(os.sep, "/")
 
 
+class RedirectEntry(TypedDict):
+    hashes: list[tuple[str, str]]
+    overall: str
+
+
+def build_redirect_entries(redirects: dict):
+    """
+    This builds a more-detailed lookup table from the original old->new page mappings.
+
+    For each old page, it contains an overall redirect of where to go,
+    as well as specific redirects for each hash, contained in a (hash, redirect) structure.
+    """
+    redirect_entries: dict[str, RedirectEntry] = {}
+    for page_old, page_new in redirects.items():
+        page_old_without_hash, old_hash = _split_hash_fragment(str(page_old))
+        if page_old_without_hash not in redirect_entries:
+            redirect_entries[page_old_without_hash] = {"hashes": [], "overall": ""}
+        if old_hash == "":
+            redirect_entries[page_old_without_hash]["overall"] = page_new
+        else:
+            redirect_entries[page_old_without_hash]["hashes"].append((old_hash, page_new))
+
+    # If a page doesn't have an overall redirect, use the first hash redirect
+    for page_old, redirect_map in redirect_entries.items():
+        if redirect_map.get("overall", "") == "":
+            redirect_entries[page_old]["overall"] = redirect_map["hashes"][0][1]
+
+    return redirect_entries
+
+
 class RedirectPlugin(BasePlugin):
     # Any options that this plugin supplies should go here.
     config_scheme = (
         ("redirect_maps", config_options.Type(dict, default={})),  # note the trailing comma
     )
 
+    redirect_entries: dict[str, RedirectEntry] = {}
```

Comment:
suggestion: Since this is not used as class variable but as an instance variable, and it's always re-assigned, we shouldn't assign default mutable value, only inform of the attribute presence.

```suggestion
    redirect_entries: dict[str, RedirectEntry]
```